event_listener/dfusion_db/models.py

```diff
@@ -135,3 +135,72 @@ def to_dictionary(self) -> Dict[str, Any]:
             "buyAmount": str(self.buy_amount),
             "sellAmount": str(self.sell_amount)
         }
+
+
+class AuctionResults(NamedTuple):
+    prices: List[int]
+    buy_amounts: List[int]
+    sell_amounts: List[int]
+
+    @classmethod
+    def from_dictionary(cls, data: Dict[str, List[int]]) -> "AuctionResults":
+        event_fields = ('prices', 'buy_amounts', 'sell_amounts')
+        assert all(k in data for k in event_fields), "Unexpected keys. Got {}".format(data.keys())
+
+        return AuctionResults(
+            data["prices"],
+            data["buy_amounts"],
+            data["sell_amounts"]
+        )
+
+
+class AuctionSettlement(NamedTuple):
+    auction_id: int
+    state_index: int
+    state_hash: str
+    prices_and_volumes: str  # Emitted as Hex String
+
+    @classmethod
+    def from_dictionary(cls, data: Dict[str, Any]) -> "AuctionSettlement":
+        event_fields = ('auctionId', 'stateIndex', 'stateHash', 'pricesAndVolumes')
+        assert all(k in data for k in event_fields), "Unexpected Event Keys: got {}".format(data.keys())
+        return AuctionSettlement(
+            int(data['auctionId']),
+            int(data['stateIndex']),
+            str(data['stateHash']),
+            str(data['pricesAndVolumes']),
+        )
+
+    def to_dictionary(self) -> Dict[str, Any]:
+        return {
+            "auctionId": self.auction_id,
+            "stateIndex": self.state_index,
+            "stateHash": self.state_hash,
+            "pricesAndVolumes": self.prices_and_volumes,
+        }
+
+    def serialize_solution(self, num_tokens: int) -> AuctionResults:
+        """Transform Byte Code for prices_and_volumes into Prices & TradeExecution objects"""
+        logging.info("Serializing Auction Results (from bytecode)")
+
+        # TODO - pass num_orders (as read from DB for solution verification)
+        tmp = self.prices_and_volumes[2:]
+        hex_str_array = [tmp[i: i+24] for i in range(0, len(tmp), 24)]
+        byte_array = list(map(lambda t: int(t, 16), hex_str_array))
+
+        prices, volumes = byte_array[:num_tokens], byte_array[num_tokens:]
+        buy_amounts = volumes[0::2]  # Even elements
```

Comment:
I believe this is incorrect. Can you check the implementation of `Solution.bytes` to see the exact encoding.

Reply:
`Solution.bytes` is currently under reconstruction. The implementation found here is based on some discussions around possibly reducing the call data sent to the contract in the event that the order slots are not all full.

The Byte String encoding of pricesAndVolumes is now interpreted as follows

```
[p_1, p_2, ..., p_T | bA_1, sA_1, bA_2, sA_2, ...] 
```

where each entry denotes the hex representation of a 96 bit integer and

- `p_i` are prices
- `bA_i` are executed buy amounts
- `sA_i` are executed sell amounts

The reason for the interweaved buy-sell amounts is so that we can treat a zero-padded byte string as equivalent to the full length in the event that not all orders are filled.

Reply:
@josojo  will push the changes to `Solution.bytes` in this PR

Reply:
So essentially, one will be able to submit a Solution for an auction that only had 50 orders, by only submitting the bytes for each of the prices and only the trade executions of those 50 orders. All the remaining bytes will be treated as zeros.

Reply:
The same is possible in the existing solution by splitting the array in half and treating the first half as buy and second half as sell orders.

Reply:
Yes I suppose that is true. Will talk to Alex, if he hasn't pushed this already.

Reply:
This is true, but will result in more operations on the Smart Contract. Since we will need to "pad" the bytes coming in with zeros (so that the sha-hash agrees).

Reply:
Right here, in the contract

https://github.com/gnosis/dex-contracts/blob/c8f452de6a2a5c4a2858df1bcabfec0494923770/contracts/SnappAuction.sol#L119

Reply:
I still don’t see how we don’t have to pad the other solution as well. The only difference I see is that the 0 bytes could be cheaper in the solution suggested here if the last orders happen to not be fulfilled. 

This is probably something that makes sense discussion synchronously.

Reply:
Yes, we have to pad it as well, the only minor advantage of our solution is that it costs a little bit less onchain. I am personally fine with any solution, we just have to converge to one solution

Reply:
Can you explain how it is cheaper with a small example?

Reply:
In your solution, we have to pad it twice - on two separate places as in our solution we only have to pad the ending. It does not really

Reply:
Small example, 

Consider an auction which settles with the 6-order retrETH example (EUR, WINE, OIL). The solution would then be 
```
[p_E, p_O, p_W, 0, ..., 0 | bA1, sA1, bA2, sA2, ..., bA6, sA6, 0, 0, ..., 0, 0]
```
In this case we would only pass `[p_E, p_O, p_W, 0, ..., 0 | bA1, sA1, bA2, sA2, ..., bA6, sA6]` to the contract (as bytes), which would then pad these bytes when storing the sha-hash of the solution information as

```
auctions[slot].shaHash = sha256(pricesAndVolumes + "BUNCH_OF_ZERO_BYTES");
```

since, I assume, the Snark requires the hash to be of all information (even the non-execution of empty order slots).